adapters/rubicon/rubicon.go

```diff
@@ -473,3 +503,174 @@ func NewRubiconAdapter(config *adapters.HTTPAdapterConfig, uri string, xuser str
+				errs = append(errs, err)
+				return nil, errs
+			}
+			request.User = &userCopy
+		}
+
+		if request.Device != nil {
+			deviceCopy := *request.Device
+			deviceExt := rubiconDeviceExt{RP: rubiconDeviceExtRP{PixelRatio: request.Device.PxRatio}}
+			deviceCopy.Ext, err = json.Marshal(&deviceExt)
+			if err != nil {
+				errs = append(errs, err)
+				return nil, errs
+			}
+			request.Device = &deviceCopy
+		}
+
+		if request.Imp[0].Video != nil {
+			videoExt := rubiconVideoExt{Skip: rubiconExt.Video.Skip, SkipDelay: rubiconExt.Video.SkipDelay, RP: rubiconVideoExtRP{SizeID: rubiconExt.Video.VideoSizeID}}
+			request.Imp[0].Video.Ext, err = json.Marshal(&videoExt)
+		} else {
+			primarySizeID, altSizeIDs, err := parseRubiconSizes(request.Imp[0].Banner.Format)
+			if err != nil {
+				errs = append(errs, err)
+				return nil, errs
+			}
+			bannerExt := rubiconBannerExt{RP: rubiconBannerExtRP{SizeID: primarySizeID, AltSizeIDs: altSizeIDs, MIME: "text/html"}}
+			request.Imp[0].Banner.Ext, err = json.Marshal(&bannerExt)
+			if err != nil {
+				errs = append(errs, err)
+				return nil, errs
+			}
+		}
+		
+		siteExt := rubiconSiteExt{RP: rubiconSiteExtRP{SiteID: rubiconExt.SiteId}}
+		pubExt := rubiconPubExt{RP: rubiconPubExtRP{AccountID: rubiconExt.AccountId}}
+
+		if request.Site != nil {
+			siteCopy := *request.Site
+			siteCopy.Ext, err = json.Marshal(&siteExt)
+			if err != nil {
+				errs = append(errs, err)
+				return nil, errs
+			}
+			siteCopy.Publisher = &openrtb.Publisher{}
+			siteCopy.Publisher.Ext, err = json.Marshal(&pubExt)
+			request.Site = &siteCopy
+		}
+		if request.App != nil {
+			appCopy := *request.App
+			appCopy.Ext, err = json.Marshal(&siteExt)
+			if err != nil {
+				errs = append(errs, err)
+				return nil, errs
+			}
+			appCopy.Publisher = &openrtb.Publisher{}
+			appCopy.Publisher.Ext, err = json.Marshal(&pubExt)
+			request.App = &appCopy
+		}
+
+		reqJSON, err := json.Marshal(request)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		requestData = append(requestData, &adapters.RequestData {
+			Method:       "POST",
+			Uri:          a.URI,
+			Body:         reqJSON,
+			Headers:      headers,
+			AuthUserName: a.XAPIUsername,
+			AuthPassword: a.XAPIPassword,
+		})
+	}
+
+	return requestData, errs
+}
+
+func (a *RubiconAdapter) MakeBids(request *openrtb.BidRequest, response *adapters.ResponseData) ([]*adapters.TypedBid, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bids := make([]*adapters.TypedBid, 0, 5)
+	bidType := openrtb_ext.BidTypeBanner
+
+	if request.Imp[0].Video != nil {
+		bidType = openrtb_ext.BidTypeVideo
```

Comment:
Is this intentional? I trust you know your server's request/responses better than I do, but it _seems_ buggy.

What if the `BidRequest` had one `Video` Imp and one `Banner`? This looks like it would mark _all_ your bids as `video` bids.

Reply:
hm. after looking into this further I'm seeing some odd behavior with MakeBids. I'm testing using two rubicon imps, one banner and one video. I'm seeing the video BidRequest both times this function is called, while the responseData is first for the banner and then for video the second time around. is this expected?

Reply:
`MakeBids` should be called once for each HTTP call. If `MakeRequests` returns two calls, then `MakeBids` will be called once with each response. If you're seeing other behavior, then something in the core auction logic  may be buggy... but make sure that `MakeRequests` is returning what you expect first.

Reply:
yeah that's how I thought it should work too. MakeRequests is returning two calls, one with a banner imp and the other with a video imp. MakeBids is called twice but I get the video imp request both times.

Reply:
I think what's happening is the request object is being updated when MakeRequests is called here:
https://github.com/prebid/prebid-server/blob/master/exchange/bidder.go#L82

since the rp adapter returns one request per imp, the request object is being replaced by the last request returned.

Reply:
I created a PR for the issue I mentioned above (https://github.com/prebid/prebid-server/pull/265). Btw this line of code is working as expected because the RP adapter creates multiple requests with one Imp each, which is why I'm always pointing to element 0 here.

Reply:
akh... I see what you mean now. That's a really great find! This is super confusing... and I do think we need to find some way to simplify things.

How do you feel about the following rule?

> Bidders must not mutate the input arguments.

There will be some performance impact... but it does have some pretty major benefits.

- It avoids this confusing mess
- It prevents subtle race conditions (see #141 and some of the other issues & PRs it links)
- We can enforce it programmatically across all Bidders (assuming something like #234 makes it through)

Reply:
I can see how useful not mutating input args would be in avoiding race conditions.
This is the situation I'm looking at. If MakeBids was called using the unmodified request object as you're suggesting, and the request object contained multiple imp objects, I guess we could loop through the request imps to match the response with the corresponding request impid. But why do all that if we can just return the request object containing the data that was actually used to make the request?

Reply:
yeah, that makes sense.

How do you feel about adding a `RequestData` param to `MakeBids`?